### PR TITLE
Fix expression labels in `guide_coloursteps()` and `guide_bins()`

### LINKED ISSUE #6005: scale_color_stepsn() will not accept expression vector for scale labels
An error is generated when attempting to specify color scale labels as a plotmath expression vector. The error is also generated when using `scale_fill_stepsn()`. I expected the color scale labels to be rendered using plotmath.

Here is code to reproduce the bug:

``` r
library(ggplot2)
data.frame(x = 1:5, y = 1:5) |>
  ggplot(aes(x = x, y = y, color = y))+
  geom_text(label = parse(text = paste0("x^", 1:5)))+
  scale_color_stepsn(breaks = 1:5,
                                  colors = c("red", "orange", "yellow","green", "blue"),
                                  labels = parse(text = paste0("x^", 1:5)))
#> Error in `vec_size()`:
#> ! `x` must be a vector, not an expression vector.
```

<sup>Created on 2024-07-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #6005.

Briefly, internally we dance around `<expression>` labels by casting them to lists of expressions when storing them in `<data.frame>`s and later unpacking such lists to expressions when handing them off to {grid}. I forgot to do this for `guide_coloursteps()` and `guide_bins()`, which have very custom code for getting the keys from the scale. This PR rectifies that.

Reprex from the issue:

``` r
devtools::load_all("~/packages/ggplot2")
#> ℹ Loading ggplot2

p <- data.frame(x = 1:5, y = 1:5) |>
  ggplot(aes(x = x, y = y, color = y))+
  geom_text(label = paste0("x^", 1:5), parse = TRUE)+
  scale_color_stepsn(
    breaks = 1:5,
    colors = c("red", "orange", "yellow","green", "blue"),
    labels = parse(text = paste0("x^", 1:5))
  )
p
```

``` r
p + guides(colour = "bins")
```

<sup>Created on 2024-07-18 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

File: R/guide-colorsteps.R [hunk 1]
```diff
@@ -114,7 +114,11 @@ GuideColoursteps <- ggproto(
     } else {
       key$.value <- breaks
     }
-    key$.label <- scale$get_labels(breaks)
+    labels <- scale$get_labels(breaks)
+    if (is.expression(labels)) {
+      labels <- as.list(labels)
+    }
```
Comment (teunbrand) [inline R/guide-colorsteps.R:120]: Wondering if it wouldn't be better if this was the responsibility of `Scale$get_labels()`.

Comment (teunbrand) [reply R/guide-colorsteps.R:120]: In some places for sure.

Comment (thomasp85) [reply R/guide-colorsteps.R:120]: I think that it logically belongs in `get_labels()` as it ensures the output of that function is usable. It shouldn't be the responsibility of the caller to further process the return value to make it work

File: R/guide-colorsteps.R [hunk 2]
```diff
@@ -111,11 +111,8 @@ GuideColoursteps <- ggproto(
     breaks <- parsed$breaks
 
     key <- data_frame0(!!aesthetic := scale$map(breaks))
-    if (even.steps) {
-      key$.value <- seq_along(breaks)
-    } else {
-      key$.value <- breaks
-    }
+    fmt <- if (even.steps) seq_along else identity
+    key$.value <- fmt(breaks)
```
[the remark this answers is not in the document]
Comment (teunbrand) [reply R/guide-colorsteps.R:115]: I'll revert this if you find the former more readable

[the remark this answers is not in the document]
Comment (thomasp85) [reply R/guide-colorsteps.R:115]: I think I do

**Comment by thomasp85** [comment]
I see you take my views on git commit messages serious :-)

**Pull request merged**